### ISSUE TITLE
Create overall project guidelines

### ISSUE THREAD

**Issue by josteink** [josteink/emacs-oob-reboot#26]
We've received lots of proposed changes, and that's great!

What seems to be lacking is a clear agreement of what kind of changes we should make, and how far reaching we should allow them to be.

Let's remember: **As such our goal is to help attract and keep new users by improving the OOB experience.**

I think we should establish some basic guidelines for what kind of changes we'd like to make, and try to be reasoned and consistent about them.

If our choices are systematic and consistent that will probably help getting them merged back upstream, especially when we need to make an argument in support of our changes. Basically it's easier to argue that "these changes *consistently* favour these and these use-cases" rather than arguing for each and every change in a seemingly random fashion.

Example of such a topic where we should strive to be consistent is the "safety features" which some users feel get in their way. Some people want them. Some people don't.

Whatever we decide (do nothing, keep the existing "safe" defaults, or disable some of them). we should aim to be consistent in our choices (without necessarily making that a slippery slope).

Quoting [my own post](https://github.com/josteink/emacs-oob-reboot/issues/24#issuecomment-299778675) in another issue:

> Consider this example: In Emacs 24.5 or Emacs 25.1, they changed prog-mode so that it by default enabled electric-brace-mode. Did you know that? Did you even notice? No you probably didn't. because you probably already had it (or electric-pair-mode) configured already, so this was effectively no change at all for you.
> 
> This change may have "broken" a few setups and configs out there (for whatever definition of broken the complainer chooses to apply), but I'm willing to bet the majority of users benefited from this change.
>
> That's the kind of settings I'd love to see in this repo.

So not to steer the conversation too much, I think our core guidelines should contain at least the following aspects:

# scope 

"Improving the OOB experience" can mean a lot of things to a lot of people. By scoping our ambitions it should be easier to propose changes, and decide when good enough is good enough.

My original aim for this project was to improve the *default behaviours* of Emacs when a user has started using it, to reduce the absolute need for "customizing" Emacs before it can be considered a good/semi-intuitive tool by modern standards.

Some people also think the *immediate* OOB experience in terms of the GUI which greets them is important to attract new users.

To what extent should we appease such users? Should we allow tweaking the GUI? Tweak the theme? *Invert* the theme to a more modern white on black setting? (And what about terminals?)

# packages

Well-established ELPA packages could be used (`company-mode` is already imported and used in `prog-mode` by a suggested change I've made), and has a chance of getting merged back in.

MELPA can't, due to cumbersome copyright-assignment issues. Or so I was told by @phillord on #Emacs on Freenode.

Are we all OK with this?

# settings

Basically any `setq` in your `.emacs`, or any other thing overriding Emacs core `defvar`s.

Are there any settings which makes Emacs-objectively better for everyone? If we can't judge it objectively, would a clear majority vote be good enough to make something good for inclusion?

# other customizations

In Emacs not all customizations are `setq`-statements.

I see some propositions where people are `defun`-ing new functions and creating keyboard bindings for them. Should we accept such changes?

Personally I lean towards the side of *not* letting such changes be included, but I'll be intellectually honest and admit I can't make a principled case for why these are different from or worse than including whole packages.

Anyone else have any strong opinions here?

**Comment by phillord** [comment]
I would work backward from the use case. Does it make the life of a new user easier to get going, to understand some of the power of Emacs without confusing. I would personally say that making it look nicer is also important; I know that Emacs' somewhat ugly interface puts people of. Not that I, personally, will be able to do better, cause my UIs are equally horrible.

In terms of customisations vs new functions, I don't care. Emacs functionality is defined by its whole code base, and I would put nothing out-of-scope. What I do care about is that we are clear that the point is to get the changes back into core. I do not want to follow the arc of starter-kit -- code to documentation guide. The implementation is unimportant, it's understanding what behaviour we want that is.

**Comment by dchrzanowski** [comment]
I completely agree with the above statements. We are here to improve the NEW user's experience of Emacs, thus, GUI is definitely important, I mean freshly opened Emacs is bloody hideous, I remember my own reaction of something alike "Which era is this from?", but after applying a single theme...boom!!! Beautiful! Maybe a few changes to the modeline as well for the cherry on top.

Regarding packages...well it is unfortunate that we cannot provide Melpa out of the box? I hear that use-package might be included in the new version of Emacs, can we add `(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/"))` or is this too much as well? Most of great editors would have access to all available packages OOB. 

Settings and others...I would be of an opinion as to only allow `setq` , however, if someone has a pretty neat and sort of game changing `defun` then why not, right? Main question to ask in this case would be: does it improve the user experience and out of the box functionality? I know that this is a simple yes/no, but opinions are strong and the struggle will be there anyway.

To sum it up, for the time being I think that we should focus on the general look of the GUI (just a nice dark theme (vote?) and a modeline modification) and those minor tweaks via `setq`.

**Comment by josteink** [comment]
> can we add `(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/"))` or is this too much as well?

As I understand it, **we can't use any MELPA packages with any workaround no matter what**, because that means that when Emacs is shipped by FSF (which is anal about copyright-assignment) it also effectively ships MELPA code the FSF don't have copyright-assignment for. And they simply won't accept that.

So no MELPA, even if that probably sounds like the strangest thing ever to pretty much every Emacs-user outside the FSF.

It's a rough constraint to work within, but I believe we should still be able to improve other parts of Emacs even with this lacking.

**Comment by josteink** [comment]
> In terms of customisations vs new functions, I don't care. ... The implementation is unimportant, it's understanding what behaviour we want that is.

If that is a sentiment shared by the majority I have no problem with that. Like I said: "I'll be intellectually honest and admit I can't make a principled case for why these are different from or worse than including whole packages."

> What I do care about is that we are clear that the point is to get the changes back into core. I do not want to follow the arc of starter-kit -- code to documentation guide.

Absolutely. This is crucial, and if the project diverts from this path, it has IMO effectively lost its entire purpose.

**Comment by angrybacon** [comment]
I'm not so hot on including default from packages that are not shipped with Emacs yet. There must be some kind of specific steps to follow to merge a package into Emacs eg. licenses, code sanity, sane defaults, active maintainers. I don't actually know, I'm only guessing so correct me if I'm wrong on that. Point is, assuming what I just said is relevant: how can a bucket of better default for Emacs take responsibility in including other packages, something that might involve more than actual code and UX choices.

TL;DR: No external packages whatsoever.

But `company` (read auto-completion at point) *is* a sane default in my opinion, so are `markdown-mode` (open a Markdown file), `json-mode` (open a JSON file) and `flycheck` (embedded linter) amongst many others I'm sure. Those are things someone should expect from any half decent text editor/IDE where the plan is to be somewhat efficient.

TL;DR: But it's too good.

It might be a too brutal to ship a repository to Emacs maintainers that includes external packages, and like I said might involve a little more than something like `(mouse-avoidance-mode 'banish)`. Why not make a list of packages that we want to include in Emacs before including these in this repository? With that idea in mind, I have no idea how long it takes for a single package to travel from MELPA to Emacs itself. That only depends on the package you might say.

**Comment by phillord** [comment]
I think we should probably add some labels to these issues -- once I can work out how to. "meta", "pretty", "frictionless", "on-by-default".

**Comment by josteink** [comment]
Thats a good idea. I'll see if I can get it done. 

Should I nuke the default labels in the process?

**Comment by angrybacon** [comment]
I think some still stand, but yeah it would be easier to pick one if there aren't too many. I'm guessing we can remove "enhancement" and "wontfix".

**Comment by josteink** [comment]
Done and done. If we need more, they can be managed on the following URL:

https://github.com/josteink/emacs-oob-reboot/labels

(And no shame in not finding them. Github kinda "hides" this URL and I always have to go looking for a long time when I need it :)

**Comment by leppert** [comment]
Thoughts on adding this issue's text, or a link to the issue, in the contribution guidelines that users see when they go to open a new issue on Github?

**Comment by josteink** [comment]
@leppert I'm for it. If someone makes a PR it will surely be merged.

**Comment by josteink** [comment]
@leppert I'm for it. If someone makes a PR it will surely be merged.

**Comment by angrybacon** [comment]
@leppert I already added a [CONTRIBUTING.org](https://github.com/josteink/emacs-oob-reboot/blob/master/CONTRIBUTING.org), feel free to add/reword things.

Edit: Nevermind, didn't see your pull request in time.
